File: workspace/platf-bizz-prjs/drr-tnt-swfabrik-europe/food-market-03-service/internal/domain/food_trading.go
package domain

import (
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

type FoodTrading struct {
	ID         int64     `json:"id"`
	TradingID  string    `json:"tradingId"`
	MarketID   string    `json:"marketId"`
	ItemName   string    `json:"itemName"`
	Quantity   float64   `json:"quantity"`
	UnitPrice  float64   `json:"unitPrice"`
	TotalPrice float64   `json:"totalPrice"`
	TraderName string    `json:"traderName"`
	Status     string    `json:"status"`
	CreatedAt  time.Time `json:"createdAt"`
}

func (f *FoodTrading) ValidateAndCalculate() error {
	if strings.TrimSpace(f.ItemName) == "" {
		return errors.New("itemName must not be blank")
	}
	if f.Quantity <= 0 {
		return errors.New("quantity must be greater than zero")
	}
	if f.UnitPrice <= 0 {
		return errors.New("unitPrice must be greater than zero")
	}
	if strings.TrimSpace(f.TraderName) == "" {
		return errors.New("traderName must not be blank")
	}
	if strings.TrimSpace(f.TradingID) == "" {
		f.TradingID = fmt.Sprintf("TRD-GO-%s", strings.ToUpper(uuid.New().String()[:8]))
	}
	if strings.TrimSpace(f.Status) == "" {
		f.Status = "CONFIRMED"
	}
	if f.CreatedAt.IsZero() {
		f.CreatedAt = time.Now().UTC()
	}
	f.TotalPrice = f.Quantity * f.UnitPrice
	return nil
}

type FoodTradingEvent struct {
	EventID    string    `json:"eventId"`
	EventType  string    `json:"eventType"`
	TradingID  string    `json:"tradingId"`
	MarketID   string    `json:"marketId"`
	ItemName   string    `json:"itemName"`
	Quantity   float64   `json:"quantity"`
	UnitPrice  float64   `json:"unitPrice"`
	TotalPrice float64   `json:"totalPrice"`
	TraderName string    `json:"traderName"`
	Status     string    `json:"status"`
	Timestamp  time.Time `json:"timestamp"`
}

type CreateFoodTradingCommand struct {
	ItemName   string  `json:"itemName"`
	Quantity   float64 `json:"quantity"`
	UnitPrice  float64 `json:"unitPrice"`
	TraderName string  `json:"traderName"`
}

File: workspace/platf-bizz-prjs/drr-tnt-swfabrik-europe/food-market-06-service/src/Domain/Entities/FoodTrading.cs
namespace Darueira.FoodMarket.Service06.Domain.Entities;

public class FoodTrading
{
    public long? Id { get; set; }
    public string TradingId { get; set; } = string.Empty;
    public string MarketId { get; set; } = string.Empty;
    public string ItemName { get; set; } = string.Empty;
    public decimal Quantity { get; set; }
    public decimal UnitPrice { get; set; }
    public decimal TotalPrice { get; set; }
    public string TraderName { get; set; } = string.Empty;
    public string Status { get; set; } = "CONFIRMED";
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public void ValidateAndCalculate()
    {
        if (string.IsNullOrWhiteSpace(ItemName))
            throw new ArgumentException("itemName must not be blank");
        if (Quantity <= 0)
            throw new ArgumentException("quantity must be greater than zero");
        if (UnitPrice <= 0)
            throw new ArgumentException("unitPrice must be greater than zero");
        if (string.IsNullOrWhiteSpace(TraderName))
            throw new ArgumentException("traderName must not be blank");
        if (string.IsNullOrWhiteSpace(TradingId))
            TradingId = $"TRD-CS-{Guid.NewGuid().ToString("N")[..8].ToUpperInvariant()}";
        if (string.IsNullOrWhiteSpace(Status))
            Status = "CONFIRMED";

        TotalPrice = decimal.Round(Quantity * UnitPrice, 2);
    }
}
